plugins/gstreamer/src/GStreamerVideoSource.cpp

```diff
@@ -0,0 +1,285 @@
+
+	gst_app_sink_set_callbacks(GST_APP_SINK(m_implementation->sink), &callbacks, this, nullptr);
+	gst_element_set_state(m_implementation->pipeline, GST_STATE_PLAYING);
+
+	m_lastError = B4BError::ok();
+}
+
+GStreamerVideoSource::~GStreamerVideoSource() noexcept
+{
+	gst_element_set_state(GST_ELEMENT(m_implementation->pipeline), GST_STATE_NULL);
+	gst_object_unref(m_implementation->sink);
+	gst_object_unref(m_implementation->userPipeline);
+	gst_object_unref(m_implementation->pipeline);
+}
+
+dto::ImageMetadata
+GStreamerVideoSource::metadata() const noexcept
+{
+	return m_frame.metadata();
+}
+
+std::error_code
+GStreamerVideoSource::nextFrame() noexcept
+{
+	if (B4BError::ok() != m_lastError)
+	{
+		return m_lastError;
+	}
+
+	const auto predicate = [this]() {
+		return frameReady();
+	};
+
+	{
+		std::unique_lock<decltype(m_mutex)> lock(m_mutex);
+
+		m_frameReadyCondition.wait(lock, predicate);
+
+		if (!m_frameReady)
+		{
+			// Not exactly an error, but not sure what to return here
+			return B4BError::genericError();
+		}
+
+		m_frameReady = false;
+		m_bufferReady = true;
+	}
+
+	m_bufferReadyCondition.notify_all();
+
+	return B4BError::ok();
+}
+
+dto::ImageView
+GStreamerVideoSource::frame() const noexcept
+{
+	return m_frame;
+}
+
+dto::ImageView
+GStreamerVideoSource::frame(std::byte*, std::size_t) const noexcept
+{
+	// Not implemented
+	return dto::ImageView();
+}
+
+std::error_code
+GStreamerVideoSource::execute(const std::string&) noexcept
+{
+	return B4BError::ok();
+}
+
+int
+GStreamerVideoSource::preroll(void*, GStreamerVideoSource* self)
+{
+	{
+		std::unique_lock<decltype(self->m_mutex)> lock(self->m_mutex);
+		self->m_bufferReady = true;
+	}
+
+	self->m_bufferReadyCondition.notify_all();
+
+	return GST_FLOW_OK;
+}
+
+int
+GStreamerVideoSource::grabFrame(void* sink, GStreamerVideoSource* self)
+{
+	const auto predicate = [self]() {
+		return self->bufferReady();
+	};
+
+	{
+		std::unique_lock<decltype(self->m_mutex)> lock(self->m_mutex);
+		self->m_bufferReadyCondition.wait(lock, predicate);
+	}
+
+	const auto appsink = static_cast<GstAppSink*>(sink);
+	const auto sample = gst_app_sink_pull_sample(appsink);
+
+	if (!sample)
+	{
+		if (gst_app_sink_is_eos(appsink))
+		{
+			{
+				std::unique_lock<decltype(self->m_mutex)> lock(self->m_mutex);
+				self->m_endOfStream = true;
+			}
+			self->m_frameReadyCondition.notify_all();
+			return GST_FLOW_OK;
+		}
+		else
+		{
+			self->m_lastError = B4BError::genericError();
+			return GST_FLOW_ERROR;
+		}
+	}
+
+	const auto caps = gst_sample_get_caps(sample);
+	const auto structure = gst_caps_get_structure(caps, 0);
+
+	auto width = 0U;
+	auto height = 0U;
+
+	// TODO
+	// Test that these actually succeed in various scenarios.
+	gst_structure_get_uint(structure, "width", &width);
+	gst_structure_get_uint(structure, "height", &height);
+
+	self->m_implementation->sample = std::make_unique<Sample>(sample);
+	self->m_frame = self->m_implementation->sample->imageView(width, height);
```

Comment:
How long is the frame valid for?

Reply:
Until `nextFrame()` is called. The grabber thing waits for the frame ready condition before getting the next sample.